### PR TITLE
Write http kairosdb format

### PR DESCRIPTION
**Pull request by beorn-**
Adds KairosDB format for write_http collectd plugin.

### CODE REVIEW

**Comment by gbolo** [comment]
Excellent new feature!

**Comment by beorn-** [comment]
been in production here for 3 monthes with ~ 60 datapoints/second without a glitch

**Comment by octo** [comment]
Thanks @beorn-! The code is very straight forward and easy to read, thank you! I'll leave some minor review comments inline. Once they're resolved, this is good to go IMO.

Best regards,
—octo

File: src/utils_format_kairosdb.c [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Aurelien beorn Rougemont <beorn at gandi dot net>
+ **/
+
+#include <ctype.h>
```
Comment (octo) [inline src/utils_format_kairosdb.c]: `#include "collectd.h"` must be the first include. Please put the `<ctype.h>` include after all the `#inlcude "…"` includes.

File: src/utils_format_kairosdb.c [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+                vl->type,
+                vl->type_instance,
+                ds->ds[ds_idx].name);
+      return (-1);
+    }
+  }
+  else if (ds->ds[ds_idx].type == DS_TYPE_COUNTER)
+  {
+    BUFFER_ADD ("[[");
+    BUFFER_ADD ("%"PRIu64, CDTIME_T_TO_MS (vl->time));
```
Comment (octo) [inline src/utils_format_kairosdb.c]: Counters should be formatted with `%llu`.

File: src/utils_format_kairosdb.c [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    BUFFER_ADD (",");
+    BUFFER_ADD ("%llu", vl->values[ds_idx].counter);
+  }
+  else if (ds->ds[ds_idx].type == DS_TYPE_DERIVE)
+  {
+    BUFFER_ADD ("[[");
+    BUFFER_ADD ("%"PRIu64, CDTIME_T_TO_MS (vl->time));
```
Comment (octo) [inline src/utils_format_kairosdb.c]: Derives should be formatted with `PRIi64`.

File: src/utils_format_kairosdb.c [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+  for (i = 0; i < ds->ds_num; i++)
+  {
+    /* All value lists have a leading comma. The first one will be replaced with
+     * a square bracket in `format_kairosdb_finalize'. */
+    BUFFER_ADD (",{");
+
+    BUFFER_ADD ("\"name\":\"collectd");
+
+    if (strlen(vl->plugin))
```
Comment (octo) [inline src/utils_format_kairosdb.c]: You can safely assume that *all* value lists have a non-empty `plugin` field.

File: src/utils_format_kairosdb.c [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+    memset (temp, 0, sizeof(temp));
+
+    BUFFER_ADD (", \"tags\":\{");
+
+    BUFFER_ADD ("\"host\": \"%s\"", vl->host);
+    if (strlen(vl->plugin_instance))
+      BUFFER_ADD_KEYVAL ("plugin_instance", vl->plugin_instance);
+    if (strlen(vl->type))
+      BUFFER_ADD_KEYVAL ("type", vl->type);
```
Comment (octo) [inline src/utils_format_kairosdb.c]: You can safely assume that *all* valud lists have a non-empty `type` field.

File: src/utils_format_kairosdb.c [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+    if (strlen(vl->type_instance))
+      BUFFER_ADD_KEYVAL ("type_instance", vl->type_instance);
+    if (ds->ds_num > 0)
```
Comment (octo) [inline src/utils_format_kairosdb.c]: All data sets have at least one data source. Did you maybe mean `ds->ds_num != 1`?

File: src/write_http.c [hunk 7]
```diff
@@ -507,6 +569,8 @@ static int config_set_format (wh_callback_t *cb, /* {{{ */
                 cb->format = WH_FORMAT_COMMAND;
         else if (strcasecmp ("JSON", string) == 0)
                 cb->format = WH_FORMAT_JSON;
+        else if (strcasecmp ("KAIROSDB", string) == 0)
```
Comment (octo) [inline src/write_http.c]: Please add this format option to the *collectd.conf(5)* manpage. It is generated from `src/collectd.conf.pod`.

Comment (beorn-) [inline src/write_http.c]: done

File: src/utils_format_kairosdb.c [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+    if (rates == NULL)
+    {
+      WARNING ("utils_format_kairosdb: uc_get_rate failed for %s|%s|%s|%s|%s",
+                vl->plugin,
+                vl->plugin_instance,
+                vl->type,
+                vl->type_instance,
+                ds->ds[ds_idx].name);
+
+      sfree(rates);
```
Comment (rubenk) [inline src/utils_format_kairosdb.c]: this has no effect since rates is always NULL here.

File: src/utils_format_kairosdb.c [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+
+#include "collectd.h"
+#include "plugin.h"
+#include "common.h"
+
+#include "utils_cache.h"
+#include "utils_format_kairosdb.h"
+
+#include <ctype.h>
```
Comment (rubenk) [inline src/utils_format_kairosdb.c]: ctype.h is already included by collectd.h

File: src/utils_format_kairosdb.c [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    }
+    else
+    {
+      WARNING ("utils_format_kairosdb: invalid rates[ds_idx] for %s|%s|%s|%s|%s",
+                vl->plugin,
+                vl->plugin_instance,
+                vl->type,
+                vl->type_instance,
+                ds->ds[ds_idx].name);
+      return (-1);
```
Comment (rubenk) [inline src/utils_format_kairosdb.c]: you leak rates here.

Comment (beorn-) [inline src/utils_format_kairosdb.c]: goot catch

File: src/utils_format_kairosdb.c (hunk 9 above)
Comment (beorn-) [inline src/utils_format_kairosdb.c]: removing it

File: src/utils_format_kairosdb.c (hunk 8 above)
Comment (beorn-) [inline src/utils_format_kairosdb.c]: fixed in ef6ff6e302c21050f27d452318c4a76abe1c0659

**Comment by beorn-** [comment]
rebasing it against master in another PR

**Comment by beorn-** [comment]
see  collectd/collectd#1809

**Pull request closed without merging**